src/Swarm/Util/UnitInterval.hs

```diff
@@ -0,0 +1,24 @@
+-- |
+-- SPDX-License-Identifier: BSD-3-Clause
+--
+-- Creation and utilities for tge unit interval
+module Swarm.Util.UnitInterval (
+  UnitInterval,
+  getValue,
+  mkInterval,
+  safeIndex,
+) where
+
+newtype UnitInterval a = UnitInterval
+  { getValue :: a
+  }
+
+-- | Guarantees that the stored interval falls on the range
+--   @[0, 1]@. It is up to clients to ensure that the promotion
+--   to this type is lossless.
+mkInterval :: (Ord a, Num a) => a -> UnitInterval a
+mkInterval = UnitInterval . max 0 . min 1
+
+safeIndex :: RealFrac a => UnitInterval a -> [b] -> b
```

Comment:
I love this!  Can you also copy the original comment arguing why this indexing operation was safe?

Reply:
Also, this should take a `NonEmpty b` instead of a `[b]`, otherwise it could crash when given an empty list.